### PR TITLE
Always resolve references and add resolve method

### PR DESCRIPTION
**Pull request by lewish**
And a few bugs I found along the way...

- Change the behaviour of ref to always try to resolve even if the dependency doesn't exist.
- This is fairly safe as there is a deterministic mapping from a node name to a target (with the exception of assertions, that should not be referenced).
- Any missing dependencies will be picked up during validation rather than as compile time errors.
- Adds a resolve method that can be used within scripts to simply map a node name to a query-able table name without adding it to dependencies.
- Deprecates hasOutput for ops, updates docs accordingly.
- Moves the dependency wildcard resolution to compilation out of validation and adds a test for it.

Although this is more flexible, the downside of all of this is that referencing an assertion or operation will not throw any warnings. We can fix that in the future by differentiating ref'd dependencies from normal dependencies in the compiled graph, enabling us to check this during validation.

### CODE REVIEW

File: api/vm/query.ts [hunk 1]
```diff
@@ -20,7 +20,8 @@ export function compile(query: string, projectDir?: string): string {
       projectDir,
       `(function() {
         const { session } = require("@dataform/core");
-        const ref = session.ref.bind(session);
+        const ref = session.resolve.bind(session);
```
Comment (BenBirt) [inline api/vm/query.ts]: `ref` no?

File: core/assertion.ts [hunk 2]
```diff
@@ -48,7 +48,11 @@ export class AssertionContext {
 
   public ref(name: string) {
     this.assertion.dependencies(name);
-    return this.assertion.session.ref(name);
+    return this.assertion.session.resolve(name);
```
Comment (BenBirt) [inline core/assertion.ts]: `return this.resolve(name);` ?

File: core/operation.ts [hunk 3]
```diff
@@ -59,7 +57,11 @@ export class OperationContext {
 
   public ref(name: string) {
     this.operation.dependencies(name);
-    return this.operation.session.ref(name);
+    return this.operation.session.resolve(name);
```
Comment (BenBirt) [inline core/operation.ts]: ditto?

File: core/session.ts [hunk 4]
```diff
@@ -167,6 +160,27 @@ export class Session {
       graphErrors: this.graphErrors
     });
 
+    // Expand node dependency wilcards.
```
Comment (BenBirt) [inline core/session.ts]: wildcards

File: core/session.ts [hunk 5]
```diff
@@ -167,6 +160,27 @@ export class Session {
...
+      compiledGraph.assertions,
+      compiledGraph.operations
+    );
+    const allNodeNames = allNodes.map(node => node.name);
+
+    allNodes.forEach(node => {
+      const uniqueDeps: { [d: string]: boolean } = {};
+      const deps = node.dependencies || [];
+      // Add non-wildcard deps normally.
+      deps.filter(d => !d.includes("*")).forEach(d => (uniqueDeps[d] = true));
```
Comment (BenBirt) [inline core/session.ts]: s/d/dep/ (throughout)

File: docs/pages/guides/operations.mdx [hunk 6]
```diff
@@ -36,14 +36,12 @@ CREATE OR REPLACE VIEW sometable.someschema AS (SELECT 1 AS TEST)
 
 ## Declaring and referencing outputs
 
-If an operation creates a table or view that you would like to make available to other scripts, then you can use the `hasOutput()` configuration method to tell
-Dataform that the output of this operation should be made available to callers of the `ref()` function.
+If an operation creates a table or view that you would like to make available to other scripts, then you can use reference this operation like you would a normal table using the `ref()` function. Additionally operations may use the `self()` to create a table or view that matches the current file name.
```
Comment (BenBirt) [inline docs/pages/guides/operations.mdx]: If an operation creates a table or view that you would like to make available to other scripts, you can reference this operation as you would a normal table by using the `ref()` function. Note that operations may use `self()` to create a table or view that matches the current file name.

File: api/vm/query.ts (hunk 1 above)
Comment (lewish) [reply api/vm/query.ts]: I renamed session.ref to session.resolve as that's all it's really doing.

**Pull request merged**